### DISCUSSION TITLE
How do I personalize (redirect) without killing performance (cache)?

### DISCUSSION THREAD

**Discussion by u840903** [vercel/next.js#19285] [Help] [answered]
Hi! 

I've got a website on Vercel running Next. I want returning users to see one start-page and anonymous newcomers another. Since these two views will differ a lot I've decided to split into two non-overlapping component structures and put them on two different urls.

- domain.com
- domain.com/home

When a returning user hits domain.com I want to redirect them to domain.com/home - an anonymous newcomer should not be redirected. How do I do this without killing performance?

**I started with an effect on domain.com.**
```
const user = getCookie('user');
useEffect(() => {
  if(user) router.push('/home');
}, []);
```
- GOOD: Excellent Time-To-First-Byte since I can cache domain.com on the Vercel Edge Network™  edge cache.
- BAD: The users sees the contents of domain.com and then domain.com/home
- BAD: Perceived performance is bad.

**Not good enough so I switched from `getStaticProps` to `getServerSideProps`.**
```
export const getServerSideProps = async ({req, res}) => {
    return req.cookies.user
        ? {redirect: {destination: "/home"}}
        : {props: {page: 'Hello Newcomers!'}}
};
```
- BAD: Bad TTFB since I dont't cache at all
- GOOD: Perceived performance is better than with the effect.
- BAD: Perceived performance is bad.

**So lets add back the cache then...**
```
export const getServerSideProps = async ({req, res}) => {
    res.setHeader('Cache-Control', 's-maxage=3600, stale-while-revalidate');
    return req.cookies.user
        ? {redirect: {destination: "/home"}}
        : {props: {page: 'Hello Newcomers!'}}
};
```
+ GOOD: Excellent Performance
- BAD: Doesn't work. (Caching one user's data and shoving it to another is not very GDPR-friendly)

I'm currently out of ideas. How would you personalise whilst keeping decent performance? My requests go through Amazon Route 53 - can it be solved there?

**Comment by rafaelalmeidatk** [comment, marked as answer]
You can add some JS directly into the the `head` tag to redirect the user if the cookie exists, since this code is going to run first than anything else the redirect is pretty fast. Here is an example: 

https://github.com/leerob/fastfeedback/blob/5e17003204484f976f08f3cafe3563fd1fbafbbe/pages/index.js#L33-L43

**Comment by u840903** [comment]
```
 <Head>
  <script
    dangerouslySetInnerHTML={{
          __html: `
            if (document.cookie && document.cookie.includes('user')) {
              window.location.href = "/home"
            }
          `
        }}
    />
</Head>
```

Thats it! 
- GOOD: Excellent TTFB on domain.com
- GOOD: Excellent TTFB on domain.com/home
- GOOD: Excellent perceived performance (The redirect from domain.com is underway within 20-50ms)

Thank you, @rafaelalmeidatk!
